numpy/core/src/multiarray/arrayobject.c

```diff
@@ -1358,18 +1358,17 @@ array_richcompare(PyArrayObject *self, PyObject *other, int cmp_op)
             _res = PyArray_CanCastTypeTo(PyArray_DESCR(self),
                                          PyArray_DESCR(array_other),
                                          NPY_EQUIV_CASTING);
+            Py_XDECREF(result);
```

Comment:
I don't think this is quite right here (or below), as a NULL return indicates that `PyArray_GenericBinaryFunction` encountered and error. Probably something like
```
if (result == NULL) {
    PyErr_Clear();
}
else {
    Py_DECREF(result);
}
```
Somehow I think the logic in this function can be cleaned up, but that might be for another day.